bpf/lib/egress_gw.h

```diff
@@ -0,0 +1,68 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+/* Copyright (C) 2019-2021 Authors of Cilium */
+
+#ifndef __EGRESS_GW_H_
+#define __EGRESS_GW_H_
+
+#include <bpf/ctx/ctx.h>
+#include <bpf/api.h>
+
+#include "eps.h"
+#include "tailcall.h"
+#include "nat.h"
+#include "edt.h"
+#include "lb.h"
+#include "common.h"
+#include "overloadable.h"
+#include "conntrack.h"
+#include "csum.h"
+#include "encap.h"
+#include "trace.h"
+
+#ifdef ENABLE_EGRESS_GATEWAY
+
+static __always_inline int handle_egress_nat_ipv4(struct __ctx_buff *ctx)
+{
+	struct egress_info *info;
+	struct iphdr *ip4;
+	void *data, *data_end;
+	bool from_endpoint = true;
+	struct ipv4_nat_target target = {
+		.min_port = 32767,
+		.max_port = 43835,
+	};
+
+	if (!revalidate_data(ctx, &data, &data_end, &ip4))
+		return DROP_INVALID;
+
+	info = lookup_ip4_egress_endpoint(ip4->saddr, ip4->daddr);
+	if (!info)
+		return CTX_ACT_OK;
+
+	target.addr = info->egress_ip;
+
+	return snat_v4_process(ctx, NAT_DIR_EGRESS, &target, from_endpoint);
```

Comment:
The egress gateway currently relies on (1) BPF NodePort and (2) tunneling mode. Could you add checks in `initKubeProxyReplacementOptions()` and `daemon/cmd/daemon_main.go` to disable the egress gateway feature with a warning if any of these two conditions isn't satisfied?

Reply:
Will add check for these 2 options: `--kube-proxy-replacement` and `--enable-bpf-masquerade`

Reply:
Sounds good :+1: